ff/src/biginteger/macros.rs

```diff
@@ -12,40 +12,71 @@ macro_rules! bigint_impl {
         impl BigInteger for $name {
             const NUM_LIMBS: usize = $num_limbs;
 
-            #[inline]
+            #[ark_ff_asm::unroll_for_loops]
             fn add_nocarry(&mut self, other: &Self) -> bool {
                 let mut carry = 0;
 
-                for (a, b) in self.0.iter_mut().zip(other.0.iter()) {
-                    *a = adc!(*a, *b, &mut carry);
+                for i in 0..$num_limbs {
+                    #[cfg(all(target_arch = "x86_64", feature = "asm"))]
+                    #[cfg_attr(all(target_arch = "x86_64", feature = "asm"), allow(unsafe_code))]
+                    unsafe {
+                        carry = core::arch::x86_64::_addcarry_u64(
+                            carry,
+                            self.0[i],
+                            other.0[i],
+                            &mut self.0[i],
+                        )
+                    };
+
+                    #[cfg(not(feature = "asm"))]
+                    {
+                        self.0[i] = adc!(self.0[i], other.0[i], &mut carry);
+                    }
                 }
 
                 carry != 0
             }
 
-            #[inline]
+            #[ark_ff_asm::unroll_for_loops]
             fn sub_noborrow(&mut self, other: &Self) -> bool {
                 let mut borrow = 0;
 
-                for (a, b) in self.0.iter_mut().zip(other.0.iter()) {
-                    *a = sbb!(*a, *b, &mut borrow);
+                for i in 0..$num_limbs {
+                    #[cfg(all(target_arch = "x86_64", feature = "asm"))]
+                    #[cfg_attr(all(target_arch = "x86_64", feature = "asm"), allow(unsafe_code))]
```

Comment:
I think this is unnecessarily verbose since it only compiles if the cfg is true. So previous version is fine.